### ISSUE TITLE
Support for triggering a azurerm_container_app_job with new resource type

### ISSUE THREAD

**Issue by jaredfholgate** [hashicorp/terraform-provider-azurerm#26885]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave comments along the lines of "+1", "me too" or "any updates", they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment and review the [contribution guide](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/contributing/README.md) to help.

### Description

I would like to be able to immediately trigger a container app job after it's creation. Currently there is no way to achieve this with Terraform.

At the moment the only fudge is to create a scheduled job and schedule for a few minutes ahead of time.

### New or Affected Resource(s)/Data Source(s)

azurerm_container_app_job_schedule_run_now

### Potential Terraform Configuration

```hcl
Taking the same pattern as `azurerm_container_registry_task_schedule_run_now`, create a new resource called `azurerm_container_app_job_schedule_run_now`.


resource "azurerm_resource_group" "example" {
  name     = "example-resources"
  location = "West Europe"
}

resource "azurerm_log_analytics_workspace" "example" {
  name                = "example-log-analytics-workspace"
  location            = azurerm_resource_group.example.location
  resource_group_name = azurerm_resource_group.example.name
  sku                 = "PerGB2018"
  retention_in_days   = 30
}

resource "azurerm_container_app_environment" "example" {
  name                       = "example-container-app-environment"
  location                   = azurerm_resource_group.example.location
  resource_group_name        = azurerm_resource_group.example.name
  log_analytics_workspace_id = azurerm_log_analytics_workspace.example.id
}

resource "azurerm_container_app_job" "example" {
  name                         = "example-container-app-job"
  location                     = azurerm_resource_group.example.location
  resource_group_name          = azurerm_resource_group.example.name
  container_app_environment_id = azurerm_container_app_environment.example.id

  replica_timeout_in_seconds = 10
  replica_retry_limit        = 10
  manual_trigger_config {
    parallelism              = 4
    replica_completion_count = 1
  }

  template {
    container {
      image = "repo/testcontainerAppsJob0:v1"
      name  = "testcontainerappsjob0"
      cpu    = 0.5
      memory = "1Gi"
    }
  }
}

resource "azurerm_container_app_job_schedule_run_now" "example" {
  container_app_job_id = azurerm_container_app_job.example.id
}
```

### References

This API should be able to trigger it: 

- https://learn.microsoft.com/en-us/rest/api/containerapps/container-apps/start?view=rest-containerapps-2024-03-01&tabs=HTTP
- https://github.com/hashicorp/go-azure-sdk/blob/main/resource-manager/containerapps/2023-05-01/jobs/method_start.go

**Comment by saliceti** [comment]
This would be extremely useful to run a database migration job (for django, rails...) before the main container app runs.
It could also be a new option in azurerm_container_app_job or a "immediate_trigger_config" block.

**Comment by rfk-nc** [comment]
Agreed this would be a really useful little resource to have access to!

**Comment by jaredfholgate** [comment]
If you need to do this, you can use AzAPI. Example here: https://github.com/Azure/terraform-azurerm-avm-ptn-cicd-agents-and-runners/blob/4002d630f190e6d5f1cdb32eb1d3c261d4c51296/modules/container-app-job/main.tf#L73

**Comment by saliceti** [comment]
Thanks. This would start the job, but there is no guarantee it completes. So I can't configure another terraform resource depending on it.
What would be ideal is a simple option like the [kubernetes_job wait_for_completion option](https://registry.terraform.io/providers/hashicorp/kubernetes/latest/docs/resources/job#wait_for_completion-1).
I had to run a bash script in the pipeline after terraform instead: [db_migrate.sh](https://github.com/NHSDigital/manage-breast-screening/blob/main/scripts/bash/db_migrate.sh)

**Comment by jaredfholgate** [comment]
That API does wait for the job to complete. It also fails if the job fails. Unless you are seeing something different in your testing?

**Comment by saliceti** [comment]
@jaredfholgate sorry I didn't have time to test until today. I tried your solution and it does start the job but in my case it returns immediately.
You can see the code here: https://github.com/NHSDigital/dtos-manage-breast-screening/blob/test-ca-job/infrastructure/terraform/main.tf#L94

Terraform apply:
```
terraform_data.job_image: Creating...
terraform_data.job_image: Creation complete after 0s [id=e0270ace-6cb0-6d79-7a31-1de84f7c392e]
azapi_resource_action.start_db_migration: Creating...
azapi_resource_action.start_db_migration: Creation complete after 3s [id=/subscriptions/67aab2fa-fcc7-49da-9dc6-cb90f0fa0628/resourceGroups/rg-manbrs-dev-uks/providers/Microsoft.App/jobs/manage-breast-screening-dbm-dev/start]

Apply complete! Resources: 2 added, 0 changed, 0 destroyed.
```

Note: `Creation complete after 3s`

But execution history show the job actually ran for 48s:
```
manage-breast-screening-dbm-dev-ejl3wgj Succeeded 48 seconds 2025-07-15 12:28:25 2025-07-15 12:29:13
```

**Comment by jaredfholgate** [comment]
> [@jaredfholgate](https://github.com/jaredfholgate) sorry I didn't have time to test until today. I tried your solution and it does start the job but in my case it returns immediately. You can see the code here: https://github.com/NHSDigital/dtos-manage-breast-screening/blob/test-ca-job/infrastructure/terraform/main.tf#L94
> 
> Terraform apply:
> 
> ```
> terraform_data.job_image: Creating...
> terraform_data.job_image: Creation complete after 0s [id=e0270ace-6cb0-6d79-7a31-1de84f7c392e]
> azapi_resource_action.start_db_migration: Creating...
> azapi_resource_action.start_db_migration: Creation complete after 3s [id=/subscriptions/67aab2fa-fcc7-49da-9dc6-cb90f0fa0628/resourceGroups/rg-manbrs-dev-uks/providers/Microsoft.App/jobs/manage-breast-screening-dbm-dev/start]
> 
> Apply complete! Resources: 2 added, 0 changed, 0 destroyed.
> ```
> 
> Note: `Creation complete after 3s`
> 
> But execution history show the job actually ran for 48s:
> 
> ```
> manage-breast-screening-dbm-dev-ejl3wgj Succeeded 48 seconds 2025-07-15 12:28:25 2025-07-15 12:29:13
> ```

Yes, sorry you are right. It just returns an error if the job fails to start successfully, so it waits for the job to start successfully, but doesn't wait for the container to exit. Apologies for the confusion.

Given a job could potentially run for many hours depending on what it is, I'm not sure if Terraform is a good solution for this use case?

I usually see a database migration being run as a separate step in the Continuous Delivery pipeline. It doesn't really fit the Infrastructure as Code pattern, it is more of an application deployment step and should be deployed and run with updates to application code? I appreciate you are probably trying to update your container revisions with IaC and that could be ok for simple use cases. But this is common decision point about how far IaC should go with regards to provisioning and updates to applications. For your use case it may be better to separate application deployment from IaC provisioning?

You probably also want to consider if you want to plan your whole infrastructure just for an application update. Normally, infrastructure updates are less frequent than application updates and it makes sense to separate them for that reason.

I obviously don't know exactly what you are doing here, so could well be missing the point, but just trying to help.
